### DISCUSSION TITLE
Sprite3D z-fighting

### DISCUSSION THREAD

**Discussion by aterim** [jnsmalm/pixi3d#164] [Q&A]
Hey there!

Is there some way to prevent z-fighting for Sprite3D?
(without disabling the depth test)

https://take.ms/xIyUN
https://take.ms/5pC2f

I know that this is not a trivial task, but maybe there are some tricks?

**Comment by jnsmalm** [comment]
Hi!

Maybe it is, not sure though :) Why do you need another solution than disabling depth test?

**Reply by aterim** [reply]
I'm planning to make a shooter with 2d sprites in 3d environment 👀

I have already developed a similar game, but used pixi-projection ([demo](https://caricature.games/demo/mkb/)).
There is no depth test and this puts a some constraint on level design. A more complex project with a similar concept can be painful to develop.

**Reply by jnsmalm** [reply]
Do you mean there is no depth test in Pixi3D? Not sure what you mean here.

**Reply by lovelle-cardoso** [reply]
@jnsmalm I am also encountering a lot of problems with z fighting and Sprite3D when the z position of the 2 Sprite3Ds are equal and the sprites are slightly overlapping. 

Is there a way to make it so pixi3D just automatically chooses one sprite to render in front of the other rather than interleaving them like the picture above? It doesn't even particularly matter to me which one it chooses as long as it just picks one to render in front of the other instead of glitching them together like it does currently.

**Reply by jnsmalm** [reply]
You can change the behaviour of this, right now by doing this:

```
// Don't read from depth buffer
app.renderer.plugins.pipeline._spriteRenderer.state.depthTest = false

// Don't write to depth buffer
app.renderer.plugins.pipeline._spriteRenderer.state.depthMask = false
```
Note that this might not work well if you are rendering other 3D meshes.

**Reply by choephix** [reply]
> way to make it so pixi3D just automatically chooses one sprite to render in front of the other rather than interleaving them like the picture above? It doesn't even particularly matter to me which one it chooses as long as it just picks one to render in front of the other instead of glitching them together like it does currently.

Similarly, is there a way to **_force_** Pixi3d to render Sprite3D instances in a specific order, effectively ignoring z depth?

For example, after turning off depth test here, the cards on the other side of the board are completely hidden behind the large surface. I assume that's because their centers are technically behind the center of the surface. It would be helpful if I could just set the surface to `zIndex = -1` etc, and for the engine to just trust me that I'm ordering things correctly via that property.

![image](https://github.com/jnsmalm/pixi3d/assets/962412/41770475-7f8d-42ea-b49a-5f752258fd24)

**Reply by jnsmalm** [reply]
Did you try to turn off depthTest and then use `renderSortOrder` on Sprite3D? https://api.pixi3d.org/classes/Sprite3D.html#renderSortOrder

**Reply by choephix** [reply]
Perfect, thanks!!
